cpp/src/arrow/ipc/dictionary.h

```diff
@@ -21,34 +21,75 @@
 
 #include <cstdint>
 #include <memory>
-#include <unordered_map>
 #include <utility>
 #include <vector>
 
-#include "arrow/memory_pool.h"
+#include "arrow/result.h"
 #include "arrow/status.h"
+#include "arrow/type_fwd.h"
 #include "arrow/util/macros.h"
 #include "arrow/util/visibility.h"
 
 namespace arrow {
+namespace ipc {
 
-class Array;
-class DataType;
-class Field;
-class RecordBatch;
+class FieldPosition {
+ public:
+  FieldPosition() : parent_(NULLPTR), index_(-1), depth_(0) {}
```

Comment:
@bkietz The idea here is to keep track of the current position in a schema without allocating a `std::vector<int>` for each visited field, only when necessary (e.g. when a dictionary is encountered). What do you think?